### PR TITLE
[post-1.4] Fix bugs when MPI proc has no DoFs

### PR DESCRIPTION
**Pull request by tjhei**
When an MPI rank has no cells and as such no DoFs, we are triggering two problems:
- Assert in SimulatorAccess<dim>::get_fe()
- handling of nullspace constraints

I am also not sure if this works with older deal.II versions, but we will let the tester tell us. :-)

### CODE REVIEW

File: source/simulator/nullspace.cc [hunk 1]
```diff
@@ -221,13 +221,6 @@ namespace aspect
 
           }
 
-#ifdef DEBUG
-      Assert(n_left_to_find == 0, ExcMessage("Error, couldn't find a velocity DoF to constrain."));
-
-      for (unsigned int d=0; d<dim; ++d)
-        Assert(vel_idx[d] != numbers::invalid_dof_index,
-               ExcMessage("Error, couldn't find a velocity DoF to constrain."));
-#endif
```
Comment (bangerth) [inline source/simulator/nullspace.cc]: The code is supposed to check for something that shouldn't happen. If you remove it, is the surrounding code able to handle the situation that isn't supposed to happen, or will we now just silently get wrong results where we previously got an error?

**Comment by bangerth** [comment]
Good detective work. OK to merge if the tests work and you are confident about the one issue.

File: source/simulator/nullspace.cc (hunk 1 above)
Comment (tjhei) [inline source/simulator/nullspace.cc]: The old logic is the following:
1. each processor find a valid dof index
2. we take the minimum over the index and constrain that DoF.
The assert checks that each processor is successful in step 1, which is impossible if you don't own any DoFs. So this check needs to go.

If you look at the new Assert below we catch the case if no processor is able to find a dof to constrain.

So yes, this should work correctly and generate wrong results.

Comment (bangerth) [inline source/simulator/nullspace.cc]: ...generate CORRECT results?

Thanks for the explanation. If I see this right, the you are using the fact that those processors who compute `vel_idx[d]==invalid_dof_index` will pass `invalid_dof_index` into `Utilities::MPI::min`. This relies on the fact that `invalid_dof_index` is an invalid number that happens to be represented by something large. This is fragile. Could we change this to
```
dealii::Utilities::MPI::min((vel_idx[d] != invalid_dof_index) ?
                                        vel_idx[d] :
                                        dof_handler.n_dofs(),   // a number larger than all DoF indices
                mpi_communicator);
```
or similar? I think that would make it clearer. You'd need to adjust the assertion in line 241 then.

Comment (tjhei) [inline source/simulator/nullspace.cc]: yes, correct results. Oops. :-)

> This relies on the fact that invalid_dof_index is an invalid number that happens to be represented by something large. 

Yes. Good point.

**Comment by tjhei** [comment]
fixed, but I think we are running into a different bug with deal 8.2.1. :-(

**Comment by bangerth** [comment]
You mean this here?
```

--------------------------------------------------------
An error occurred in line <723> of file </root/deal.II/dealii/source/lac/trilinos_sparsity_pattern.cc> in function
    void dealii::TrilinosWrappers::SparsityPattern::compress()
The violated condition was: 
    nonlocal_graph->IndicesAreGlobal() == true
The name and call sequence of the exception was:
    ExcInternalError()
Additional Information: 
(none)

Stacktrace:
-----------
#0  /root/deal.II/installed/lib/libdeal_II.g.so.8.2.0: dealii::TrilinosWrappers::SparsityPattern::compress()
#1  /root/deal.II/installed/lib/libdeal_II.g.so.8.2.0: dealii::BlockSparsityPatternBase<dealii::TrilinosWrappers::SparsityPattern>::compress()
#2  /aspect-tester/build-gcc/tests/../aspect: aspect::Simulator<2>::setup_system_matrix(std::vector<dealii::IndexSet, std::allocator<dealii::IndexSet> > const&)
#3  /aspect-tester/build-gcc/tests/../aspect: aspect::Simulator<2>::setup_dofs()
#4  /aspect-tester/build-gcc/tests/../aspect: aspect::Simulator<2>::run()
#5  /aspect-tester/build-gcc/tests/../aspect: main
--------------------------------------------------------
```

This looks like a bug in deal.II. Does this work with deal.II 8.4? If so, I'm inclined to merge the PR anyway, simply because it is strictly forward progress: no previous test now fails, and something that failed before (but had no test) now makes it farther than it did before.

**Comment by tjhei** [comment]
Yes, it works with 8.4 and I also think it does with 8.3.0. The problem with merging this is that from then on this test will always fail.

**Comment by bangerth** [comment]
Yes, I know. Should we postpone till after 8.4 is released?

**Comment by bangerth** [comment]
(I meant ASPECT 1.4, of course.)

**Comment by tjhei** [comment]
ready to be merged!

**Pull request merged**